File: sample.py
import math
import numpy as np

#@title Funciones de generacion de muestras

#Como movimiento geometrico browneano
def generate_new_sample(sample,dt,sigma,r):
  sq_dt = math.sqrt(dt)
  Z = np.random.normal(loc=0,scale=sq_dt)
  new_sample = sample*math.exp((r - (sigma**2)/2) * dt + sigma*Z)
  return new_sample

def generate_negative_sample(prev_sample,sample,dt,sigma,r):
  exponential_term = sample / prev_sample
  divisor = exponential_term / math.exp((r - (sigma**2)/2) * dt) # e**(sigma*Z)
  negative_sample = sample / divisor**2  # sample* e**((r - (sigma**2)/2) * dt + sigma*Z - 2*sigma*Z) = sample* e**((r - (sigma**2)/2) * dt + sigma*(-Z))
  return negative_sample

#Para testear
def generate_dud_sample(sample,dt,sigma,r):
  return sample * 2

File: cpp/broadie-glasserman/sample.cc
#include <iostream>
#include <cmath>
#include <random>
#include <vector>
#include <chrono>

using namespace std;

double generate_new_sample(double sample, double dt, double sigma, double r, mt19937 &gen, normal_distribution<> &d)
{
    double Z = d(gen);
    double new_sample = sample * exp((r - (sigma * sigma) / 2) * dt + sigma * Z);
    return new_sample;
}

double generate_negative_sample(double prev_sample, double sample, double dt, double sigma, double r)
{
    double exponential_term = sample / prev_sample;
    double divisor = exponential_term / exp((r - (sigma * sigma) / 2) * dt);
    double negative_sample = sample / (divisor * divisor);
    return negative_sample;
}
